source/v5/mqtt5_to_mqtt3_adapter.c

```diff
@@ -2198,21 +2200,19 @@ static void s_aws_mqtt5_to_mqtt3_adapter_subscribe_completion_fn(
         struct aws_byte_cursor topic_filter;
         AWS_ZERO_STRUCT(topic_filter);
 
-        enum aws_mqtt_qos granted_qos = AWS_MQTT_QOS_AT_MOST_ONCE;
+        enum aws_mqtt_qos granted_qos = AWS_MQTT_QOS_FAILURE;
 
-        size_t subscription_count = aws_array_list_length(&subscribe_op->subscriptions);
         if (subscription_count > 0) {
             aws_array_list_get_at(&subscribe_op->subscriptions, &record, 0);
             topic_filter = record->subscription_view.topic_filter;
         }
 
-        if (suback != NULL) {
-            if (suback->reason_code_count > 0) {
-                granted_qos = s_convert_mqtt5_suback_reason_code_to_mqtt3_granted_qos(suback->reason_codes[0]);
-            }
-        } else {
-            granted_qos = AWS_MQTT_QOS_FAILURE;
+        if (suback != NULL && suback->reason_code_count > 0) {
+            granted_qos = s_convert_mqtt5_suback_reason_code_to_mqtt3_granted_qos(suback->reason_codes[0]);
+        } else if (record) {
+            granted_qos = (enum aws_mqtt_qos)(record->subscription_view.qos);
```

Comment:
Looking at this again, I think this is not what we want to do.  This is a bug in the 311 client (to report the requested qos rather than the returned qos).  We should fix the bug instead of making the 5 adapater also do the wrong thing.

This would apply to both the single and multi acks.